### ISSUE TITLE
Fit Generator different checkpoint metric results using same generator for training and validation

### ISSUE THREAD

**Issue by ririya** [keras-team/keras#10878]
I'm wondering why I get different checkpoint metrics using the same generator for training and validation. I'm using keras version 2.0.9 with tensorflow 1.4.1. I know using the same generator does not make much sense, but I started doing this for debugging and got intrigued by this issue.

`history = m.fit_generator(G, steps_per_epoch=len(params.trainImageList) / params.train_batch_size, validation_data=G,
                              validation_steps=len(params.testImageList) / params.val_batch_size, epochs=params.epochs, callbacks=[checkpoint])
`

The losses and metrics results are very different (pdiff for example there is a 2 fold difference). Over the next epoch, the training results remain similar as well as the validation results. This is a very unusual behavior to me, I don't know if I can rely on those metrics. 

![image](https://user-images.githubusercontent.com/23408930/43931917-fc65b0d0-9bf5-11e8-8491-4c11fee82b1a.png)

**Comment by brge17** [comment]
1. The network is changing (probably improving) during the epoch. Every batch during training has a different set of weights, whereas the validation data is computed against fixed weights.

Think about it in the limiting case where your batch size is the number of data points. You compute the loss and metrics on the training set, backprop update weights, and compute the loss and metrics on the validation set, repeat... The weights have changed so the results should not be the same.

2. Unless you use stateful metrics, metrics are batch averaged. For things like False Positives etc..., the batch averaged version of two random draws from the same sequence is unlikely to produce the same result.

**Comment by ririya** [comment]
@brge17 I do expect them to be slightly different because the weights were changing throughout the epoch and the validation is calculated with fixed weights. But I don't expect the difference to be this large. What I find strange is that the results for each mini batch during training are consistent (around 0.2) and over the next epoch they have decreased a little but still around 0.18. There is a two fold difference over the validation results (around 0.1).
